### PR TITLE
Add article about Plugin::LogReport and Logger::LogReport

### PR DESCRIPTION
**Pull request by abeverley**
I promised I would write this and others for the advent calendar. Although that event has passed, the text will be useful for people anyway, and can maybe form part of next year's.

### CODE REVIEW

**Comment by xsawyerx** [comment]
Thank you for the entry, Andrew. I will be reviewing it for this year's calendar.

**Comment by abeverley** [comment]
Thanks @xsawyerx . I've added a second follow-up article about translations (ideally it should be linked to the first one where the relevant text is in the introduction). I've made a bit of a hash with the commits by doing it all on master though, so let me know if you want me to change it in any way.

File: 2015/9-easy-exceptions-and-logging.pod [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+
+If a fatal error occurs that did not originate as a L<Log::Report>
+message, then at some point it has to be translated into its L<Log::Report>
+equivalent. There is a bit of guesswork that goes into this (see the
+L<Log::Report> documentation), and sometimes it might not quite be what
+you want. For example, if the code in the above example C<die>s due to
+a coding error such as calling a method on an undefined value, then
+it will be reported as an C<ERROR>, rather than C<PANIC>. The former
+will display the message to the user, which is not what we want. I'm
+considering changing the code to report all such errors at the C<PANIC> level, as they
```
Comment (xsawyerx) [inline 2015/9-easy-exceptions-and-logging.pod]: Instead of "I'm considering ...", you can simply say "This might be changes in a future version".

File: 2015/9-easy-exceptions-and-logging.pod [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+are normally always unexpected application exceptions.
+
+
+=head2 Dispatchers
+
+I have purposefully tried to keep this article simple, without going into
+too much of the terminology of L<Log::Report>. There is one thing I should
+mention though, and that is the concept of dispatchers. A dispatcher is
```
Comment (xsawyerx) [inline 2015/9-easy-exceptions-and-logging.pod]: I would rephrase this as:

"While you can use everything so far the way I have demonstrated, there is added value in understanding the concept of `Log::Report` dispatchers."

File: 2015/9-easy-exceptions-and-logging.pod [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+something that processes a message, and can be thought of a bit like
```
Comment (xsawyerx) [inline 2015/9-easy-exceptions-and-logging.pod]: "that processes a message, similar to a Dancer logging engine".

File: 2015/9-easy-exceptions-and-logging.pod [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+           flags: "pid ndelay nowait"
+           mode: DEBUG
+
+Dispatchers can also be used to format messages.
+
+=head2 There is a lot more
+
+The above is just a flavour of the common and easy stuff you might like
+to do. C<Log::Report> itself is way, way more powerful than that though.
+I'll hopefully write a follow-up article in the future on some of the
```
Comment (xsawyerx) [inline 2015/9-easy-exceptions-and-logging.pod]: "Some of the things that can be covered in a future article..."

Generally I would not say this, because then a person wonders whether that article exists. That's something to be determined before publishing it, obviously. I would instead say "There are even more advanced topic, such as...".

If you write the articles (yes, please) and we decide to publish them, you could add "Stay tuned for additional articles on these topics during this Advent Calendar" or something like that.

File: 2015/9-easy-exceptions-and-logging.pod [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+
+=item * Add messages to a class
+
+You can add messages to a particular type of class. I use this to tag
+some as HTML content. By default, I HTML-encode all messages in a template,
+this class can be used to bypass that.
+
+=item * Use syslog for email notifications
+
+I use a Syslog dispatcher, and configure rsyslogd to email me on
```
Comment (xsawyerx) [inline 2015/9-easy-exceptions-and-logging.pod]: Link to "syslog", link to "rsyslog".?

File: 2015/9-easy-exceptions-and-logging.pod [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+any message above a certain severity. This means that I am normally
+fixing an application error before it has been reported by the user.
+
+=item * Translate message into other languages
+
+You can translate your messages into other languages using L<Log::Report::Translator>.
```
Comment (xsawyerx) [inline 2015/9-easy-exceptions-and-logging.pod]: You can mention to stay tuned for an article on it or something.

File: 2015/10-dancer-i18n-with-log-report.pod [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+More specifically, I will use its Dancer2 plugin,
+L<Dancer2::Plugin::LogReport>. If you haven't already, please look at my other
+article on that.
+
+=head2 A basic application
+
+Let's get started by creating a basic application with some strings that we want
+translated. We'll look at translating the content of templates later.
+
+ # In your route handler
```
Comment (xsawyerx) [inline 2015/10-dancer-i18n-with-log-report.pod]: Can you please indent all of the code examples to 4 spaces?

File: 2015/10-dancer-i18n-with-log-report.pod [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+
+By running this application, you should see C<Hello world> on the index page.
+
+=head2 Translating strings
+
+L<Log::Report> has L<a lot of functions to translate
+strings|Log::Report#Language-Translations>. We'll touch on 2 of the commonly
+used ones here.
+
+Firstly to simply translate a fixed string into another, we can use C<__> (2
```
Comment (xsawyerx) [inline 2015/10-dancer-i18n-with-log-report.pod]: "First, ..."

File: 2015/10-dancer-i18n-with-log-report.pod [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+
+ notice __"Hello world";
+
+This might look a bit strange, but the 2 underscores are simply a function. You
+can wrap the text in parentheses if you prefer:
+
+ notice __("Hello world");
+
+They work like any other function, although be aware that you can't use single
+quotes, as C<__'> is interpreted as C<::> by Perl
```
Comment (xsawyerx) [inline 2015/10-dancer-i18n-with-log-report.pod]: This sentence is not clear.

File: 2015/10-dancer-i18n-with-log-report.pod [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+
+We can also use C<__x>, which is the same as C<__>, except that it allows for
+variables within the string. For example:
+
+ notice __x"Hello {name}", name => 'Andy';
+
+This means that when the text is extracted as message IDs, that any variable
```
Comment (xsawyerx) [inline 2015/10-dancer-i18n-with-log-report.pod]: Remove second "that" at the end of the sentence.

File: 2015/10-dancer-i18n-with-log-report.pod [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+such as adjusting the text for pluralisation depending on variable conditions,
+but I won't go into that just yet.
+
+So, for now, update your web application accordingly:
+
+ get '/' => sub {
+   notice __"Hello world";
+   notice __x"Hello {name}", name => 'Andy';
+   template 'index';
+ };
```
Comment (xsawyerx) [inline 2015/10-dancer-i18n-with-log-report.pod]: I recommend `__(...)` and `__x(...)` to make it easier for people to read than `__...`.

File: 2015/10-dancer-i18n-with-log-report.pod [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+
+If you hadn't realised, the beauty of the above approach, is that initially you
```
Comment (xsawyerx) [inline 2015/10-dancer-i18n-with-log-report.pod]: "If you hadn't realised" sounds a bit condescending.

File: 2015/10-dancer-i18n-with-log-report.pod [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+can just add (very easily) the translation functions to all your strings. Then,
+at a later date, you can add the translation framework.
+
+=head2 Extracting and translating the strings
+
+Let's add a translater to the application, but first, we'll need some
+translation tables.
```
Comment (xsawyerx) [inline 2015/10-dancer-i18n-with-log-report.pod]: "Let's ... , but first, let's ..." - A bit confusing.

Either "We will need some translation tables in order to translate the application" or "Let's add some translation tables". You can add, "They will serve us when we want to add a translator to the application".

File: 2015/10-dancer-i18n-with-log-report.pod [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+
+=head3 Extracting and creating PO translation tables
+
+These can be extracted using L<xgettext-perl>. From the base directory of your
+application:
+
+ xgettext-perl -p messages --from-code=iso-8859-1 .
+
+This will scan the current directory for appropriate perl files and create a
```
Comment (xsawyerx) [inline 2015/10-dancer-i18n-with-log-report.pod]: "Perl", not "perl".

File: 2015/10-dancer-i18n-with-log-report.pod [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+   $msg->toString;
+ }
+ 
+ hook before_template => sub {
+   shift->{loc}  = \&translate;
+ };
+
+The above code adds a subroutine C<loc> that can be called from within each
+template. Note that we are using a different domain. This is normally best so
+as to keep the sets of messages completely separate, but it's also possible to
```
Comment (xsawyerx) [inline 2015/10-dancer-i18n-with-log-report.pod]: We have many readers who are not native English speakers. I would avoid an expression like "so as to".

"This is normally best  as to keep the sets of messages..." -> "This is a good practice in order to keep...", maybe?

File: 2015/10-dancer-i18n-with-log-report.pod [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+use the same domain if you prefer.
+
+ As a simple example, add the following to your index template:
```
Comment (xsawyerx) [inline 2015/10-dancer-i18n-with-log-report.pod]: `C<index>`.

File: 2015/10-dancer-i18n-with-log-report.pod [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+   )
+ );
+
+Re-run the application, and both the previous string as well as new template
+text should be translated.
+
+=head2 Handling plurals
+
+I won't go into the full power of L<Log::Report>, but to give a flavour of
+what's possible, here's how to handle pluralisation:
```
Comment (xsawyerx) [inline 2015/10-dancer-i18n-with-log-report.pod]: "Give a flavour" is repeated three times in both articles, and in the end at both articles. You might want to rephrase some to avoid repetition.

"Here is an example of pluralisation
[...]
(You can read more about it here.)"

Might be easier.

File: 2015/9-easy-exceptions-and-logging.pod [hunk 18]
```diff
@@ -0,0 +1,293 @@
...
+
+=item * Found yourself writing the same code in every application in order to send messages to a template
+
+=item * Wanted a clean way to handle 500 errors, whilst getting notified as the system administrator
+
+=back
+
+L<Dancer2::Plugin::LogReport> combined with L<Dancer2::Logger::LogReport>
+are designed to do all of the above, and more.
+Although the 2 modules can be used separately, it's recommended to use
```
Comment (datamuc) [inline 2015/9-easy-exceptions-and-logging.pod]: Also

    =item *

    item body

    =item *

    next item body

Comment (xsawyerx) [inline 2015/9-easy-exceptions-and-logging.pod]: I don't follow this comment. Could you expand?

File: 2015/10-dancer-i18n-with-log-report.pod (hunk 8 above)
Comment (SysPete) [inline 2015/10-dancer-i18n-with-log-report.pod]: I think I'd just start with: To translate...

But I tend to be overly-terse so perhaps ignore me. :smile:

**Comment by xsawyerx** [comment]
@abeverley, Any update? I would love to have your articles in this year and we're cutting close at this point.

**Comment by abeverley** [comment]
Sorry for the delay @xsawyerx and thanks for the feedback - I'll aim to rework tonight.

**Comment by abeverley** [comment]
Sorry @xsawyerx I've failed tonight, will aim to get on the case tomorrow.

File: 2015/9-easy-exceptions-and-logging.pod (hunk 18 above)
Comment (abeverley) [inline 2015/9-easy-exceptions-and-logging.pod]: I think it's actually meant to be one paragraph, so I've removed the line break altogether.

File: 2015/9-easy-exceptions-and-logging.pod [hunk 19]
```diff
@@ -0,0 +1,293 @@
...
+
+So, what does this give us as standard? Let's have a look:
+
+=over 4
+
+=item * Default console logging
+
+Without any configuration, the default logging is unchanged. That is, messages
+will continue to be sent to the console via STDERR (the formatting has changed,
+but we'll worry about that later). So, not really any difference there,
```
Comment (abeverley) [inline 2015/9-easy-exceptions-and-logging.pod]: Thanks, that's better. I've worded slightly differently, as the formatting is the thing we "worry" about later, and isn't actually covered in this intro (maybe it should be)

File: 2015/9-easy-exceptions-and-logging.pod [hunk 20]
```diff
@@ -0,0 +1,293 @@
...
+=item * Additional logging keywords
+
+We have some additional logging levels. Given that Dancer already provides us
+with 5, you might think this unnecessary. The extra levels give us a really handy
+feature though: we can differentiate between messages that we would normally want
+to send to the user, and messages that we would normally want sent to the system. For
+example, we can now do:
+
+ # Send a routine notice to the end user
+ notice "Hey user, I though you might want to know this"
```
Comment (abeverley) [inline 2015/9-easy-exceptions-and-logging.pod]: Fixed

File: 2015/9-easy-exceptions-and-logging.pod [hunk 21]
```diff
@@ -0,0 +1,293 @@
...
+   <div class="alert alert-[% msg.bootstrap_color %]">
+     [% msg.toString | html_entity %]
+   </div>
+ [% END %]
+
+=item * Using the message keywords without using C<Dancer2>
+
+In any other module, you can now include C<Log::Report> and use the same
+keywords/functions to send messages to your application (with the exception
+of C<success> - although there is a workaround for that if needed).
```
Comment (abeverley) [inline 2015/9-easy-exceptions-and-logging.pod]: I've worded better now.

File: 2015/9-easy-exceptions-and-logging.pod [hunk 22]
```diff
@@ -0,0 +1,293 @@
...
+=item * Clean exception handling
+
+The module will also, by default, handle any application exceptions cleanly.
+Instead of sending a 500 error, it will forward to an appropriate page, with
+the message rendered in the template as above. More on this to follow later.
+
+If the application is running with C<show_errors> enabled, then the full
+error will be sent in the message. If C<show_errors> is disabled (such as
+in a production environment), a generic error message is shown (the full
+error will have been sent to the console, or whichever backend is configured).
```
Comment (abeverley) [inline 2015/9-easy-exceptions-and-logging.pod]: Fixed.

File: 2015/9-easy-exceptions-and-logging.pod [hunk 23]
```diff
@@ -0,0 +1,293 @@
...
+
+=back
+
+So in summary, by including a couple of modules, we can easily send messages to
```
Comment (abeverley) [inline 2015/9-easy-exceptions-and-logging.pod]: Fixed. I have a habit of writing in slang ;-)

File: 2015/9-easy-exceptions-and-logging.pod [hunk 24]
```diff
@@ -0,0 +1,293 @@
...
+
+ post '/settings' => sub {
+   my $message;
+   try {
+     MyApp::Settings->update($settings);
+     $message = "Settings updated successfully";
+   }
+   catch {
+     $message = $@;
+   };
```
Comment (abeverley) [inline 2015/9-easy-exceptions-and-logging.pod]: Fixed

File: 2015/9-easy-exceptions-and-logging.pod [hunk 25]
```diff
@@ -0,0 +1,293 @@
...
+ };
+ 
+ package MyApp::Settings;
+
+ sub update {
+   ...
+   die "Invalid email address" unless ...
+ }
```
Comment (abeverley) [inline 2015/9-easy-exceptions-and-logging.pod]: Fixed

File: 2015/9-easy-exceptions-and-logging.pod [hunk 26]
```diff
@@ -0,0 +1,293 @@
...
+
+How would we do that with L<Dancer2::Plugin::LogReport>?
+
+ package MyApp;
+
+ post '/settings' => sub {
+   if (process sub{ MyApp::Settings::update($settings) }) {
+     success "Settings updated successfully";
+   }
+ };
```
Comment (abeverley) [inline 2015/9-easy-exceptions-and-logging.pod]: Actually, for this example, that could be it (hence the beauty of the plugin - no code required to handle the error message, it will already have been done). I've added that as a comment to the code.

File: 2015/9-easy-exceptions-and-logging.pod [hunk 27]
```diff
@@ -0,0 +1,293 @@
...
+
+ package MyApp::Settings;
+
+ use Log::Report;
+ sub update {
+   ...
+   error "Invalid email address" unless ...
+ }
```
Comment (abeverley) [inline 2015/9-easy-exceptions-and-logging.pod]: Fixed

File: 2015/9-easy-exceptions-and-logging.pod [hunk 28]
```diff
@@ -0,0 +1,293 @@
...
+That's it. But what if it wasn't successful? Well, the module handles all that
+for you. The C<error> function in the module is fatal just like C<die>, but as
+we showed above, its text will have automatically been sent to the template
+for display to the user.
+
+If you use C<error> outside of C<process>, then it is still fatal and the
+text is still sent to the template, but it's not handled quite as cleanly:
+the user is forwarded back to the same page if it was a C<POST> request,
+forwarded to C<'/'> if it was a C<GET> request, or fails uncleanly if it
+was already a C<GET> request to C<'/'>.
```
Comment (abeverley) [inline 2015/9-easy-exceptions-and-logging.pod]: I assume not as well, not quite sure why I put that there

File: 2015/9-easy-exceptions-and-logging.pod [hunk 29]
```diff
@@ -0,0 +1,293 @@
...
+
+As an aside, you would probably want to use the same route handler for both
+C<GET> and C<POST>. If the C<POST> is unsuccessful, the error is caught, and
+the page will behave as a C<GET> request (but with the error text rendered).
```
Comment (abeverley) [inline 2015/9-easy-exceptions-and-logging.pod]: Agreed.

File: 2015/9-easy-exceptions-and-logging.pod (hunk 2 above)
Comment (abeverley) [inline 2015/9-easy-exceptions-and-logging.pod]: Updated

File: 2015/9-easy-exceptions-and-logging.pod (hunk 4 above)
Comment (abeverley) [inline 2015/9-easy-exceptions-and-logging.pod]: Fixed, and I've mentioned the second article on translations.

File: 2015/9-easy-exceptions-and-logging.pod (hunk 5 above)
Comment (abeverley) [inline 2015/9-easy-exceptions-and-logging.pod]: Added

File: 2015/9-easy-exceptions-and-logging.pod (hunk 6 above)
Comment (abeverley) [inline 2015/9-easy-exceptions-and-logging.pod]: Done

**Comment by abeverley** [comment]
I've done the first article, will do the other tomorrow.

File: 2015/10-dancer-i18n-with-log-report.pod (hunk 7 above)
Comment (abeverley) [inline 2015/10-dancer-i18n-with-log-report.pod]: Done.

File: 2015/10-dancer-i18n-with-log-report.pod (hunk 8 above)
Comment (abeverley) [inline 2015/10-dancer-i18n-with-log-report.pod]: Ooh, the choice. I went with @SysPete's suggestion in the end - I tend to be overly slangy.

File: 2015/10-dancer-i18n-with-log-report.pod (hunk 9 above)
Comment (abeverley) [inline 2015/10-dancer-i18n-with-log-report.pod]: I've rewritten that section; hopefully it makes sense now.

File: 2015/10-dancer-i18n-with-log-report.pod (hunk 10 above)
Comment (abeverley) [inline 2015/10-dancer-i18n-with-log-report.pod]: Done.

File: 2015/10-dancer-i18n-with-log-report.pod (hunk 11 above)
Comment (abeverley) [inline 2015/10-dancer-i18n-with-log-report.pod]: I'd actually prefer to leave it as it is if possible, because it's the standard way that both Log::Report and Locale::TextDomain encourage. It makes the code more readable, especially when there is a large quantity of text to be translated (hence the reason why such a short unobtrusive function name was chosen by the modules).

File: 2015/10-dancer-i18n-with-log-report.pod (hunk 12 above)
Comment (abeverley) [inline 2015/10-dancer-i18n-with-log-report.pod]: Point taken, removed.

File: 2015/10-dancer-i18n-with-log-report.pod (hunk 16 above)
Comment (abeverley) [inline 2015/10-dancer-i18n-with-log-report.pod]: Fixed

**Comment by abeverley** [comment]
All done now, let me know if there is anything else that needs changing.

**Comment by abeverley** [comment]
Thanks for the comprehensive review.

**Comment by xsawyerx** [comment]
Thank you for these great articles and for the continued work on it. :)

I'll be merging them soon enough.

**Comment by xsawyerx** [comment]
Oh, last item: Copyright. I will be happy to add the "Author" head at the end, but we will need copyright notice. (An  article could be translated and published elsewhere, for example.)

The copyright notice I use (which I cannot vouch for its legality) is this one, also used by @SysPete in his new article:

https://github.com/PerlDancer/advent-calendar/pull/31/files#diff-c3280f970bde7de78ff29f98dffb1eedR104

**Comment by abeverley** [comment]
Brilliant. Very happy with the suggested copyright notice.

**Comment by xsawyerx** [comment]
Merged! Thank you. :)
